src/main/java/is/citizen/sdk/crypto/CitizenCrypto.java

```diff
@@ -0,0 +1,364 @@
+    }
+
+    public static Optional<KeyHolder> generateAuthenticationKeyPair(String secret) {
+        KeyHolder keyHolder = null;
+        try {
+            ECNamedCurveParameterSpec ecNamedCurveParameterSpec = ECNamedCurveTable.getParameterSpec(AUTH_CURVE);
+            KeyPairGenerator keyGen = KeyPairGenerator.getInstance(AUTH_ALGORITHM, BOUNTY_CASTLE_PROVIDER);
+            keyGen.initialize(ecNamedCurveParameterSpec);
+            KeyPair keyPair = keyGen.generateKeyPair();
+
+            String privateKey = encryptPrivateKey(keyPair.getPrivate(), secret);
+
+            String publicKey = getKeyAsString(keyPair.getPublic());
+
+            keyHolder = new KeyHolder();
+
+            keyHolder.setPrivateKey(privateKey);
+            keyHolder.setPublicKey(publicKey);
+
+        } catch (IOException | NoSuchAlgorithmException | NoSuchProviderException | InvalidAlgorithmParameterException e) {
+            throw new CryptoException("CitizenCrypto exception: " + e.getMessage());
+        }
+
+        return Optional.ofNullable(keyHolder);
+    }
+
+    public static Optional<KeyHolder> generateSigningKeyPair(String secret) {
+        return generateAuthenticationKeyPair(secret);
+    }
+
+    public static String signData(byte[] dataToSign, String encodedPrivateKey, String password) {
+		try {
+            PrivateKey privateKey = getPrivateKeyFromEncodedEncryptedString(encodedPrivateKey, password);
+			Signature signature = Signature.getInstance(AUTH_SIGNATURE_ALGORITHM);
+			signature.initSign(privateKey);
+			signature.update(dataToSign);
+			byte[] signedText = signature.sign();
+			return Base64.getEncoder().encodeToString(signedText);
+		} catch (IllegalArgumentException | NoSuchAlgorithmException | InvalidKeyException |
+                 IOException | SignatureException e) {
+            throw new CryptoException(e.getMessage());
+		}
+    }
+
+    public static boolean verifySignature(byte[] data, byte[] signedData, String encodedPublicKey)
+            throws CryptoException {
+        try {
+            KeyFactory keyFactory = KeyFactory.getInstance("EC");
+            X509EncodedKeySpec publicKeySpec = new X509EncodedKeySpec(Base64.getDecoder().decode(encodedPublicKey));
+            ECPublicKey publicKey = (ECPublicKey) keyFactory.generatePublic(publicKeySpec);
+
+            Signature verificationFunction = Signature.getInstance(AUTH_SIGNATURE_ALGORITHM);
+            verificationFunction.initVerify(publicKey);
+            verificationFunction.update(data);
+
+             if (verificationFunction.verify(signedData)) {
+                 return true;
+             }
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException | InvalidKeyException |
+                 SignatureException e)
+        {
+            throw new CryptoException("Unable to verify signature: " + e.getMessage());
+        }
+
+        return false;
+    }
+
+    public Optional<Map> encryptMap(Map map, String publicKey)
```

Comment:
We should type maps. We seem to treat both the map we pass in and the one we return as <String, String> so that's how they should be typed

Reply:
Fixed - they're now typed to <String, String>